File: Catalyst-Mono/Catalyst-DiscordCore/src/main/java/net/kingchev/catalyst/ru/command/model/AbstractCommand.java
package net.kingchev.catalyst.ru.command.model;

import net.dv8tion.jda.api.entities.Guild;
import net.dv8tion.jda.api.entities.Member;
import net.dv8tion.jda.api.entities.User;
import net.dv8tion.jda.api.events.message.MessageReceivedEvent;
import net.kingchev.catalyst.ru.command.service.InternalCommandService;
import net.kingchev.catalyst.ru.config.CommonProperties;
import net.kingchev.catalyst.ru.shared.service.DiscordEntityAccessor;
import net.kingchev.catalyst.ru.shared.service.DiscordService;
import net.kingchev.catalyst.ru.utils.TimeSequenceParser;
import org.apache.commons.lang3.StringUtils;
import org.apache.commons.lang3.tuple.Pair;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.context.ApplicationContext;

public abstract class AbstractCommand implements Command {
    @Autowired
    protected DiscordService discordService;

    @Autowired
    protected CommonProperties commonProperties;

    @Autowired
    protected InternalCommandService commandsService;

    @Autowired
    protected ApplicationContext applicationContext;

    @Autowired
    protected DiscordEntityAccessor entityAccessor;

    private CatalystCommand annotation;

    @Override
    public boolean isAvailable(User user, Guild guild, Member member) {
        return true;
    }

    protected boolean ok(MessageReceivedEvent message) {
        commandsService.resultEmotion(message, "✅", null);
        return true;
    }

    protected boolean fail(MessageReceivedEvent message) {
        commandsService.resultEmotion(message, "❌", null);
        return false;
    }

    protected boolean ok(MessageReceivedEvent message, String messageCode, Object... args) {
        commandsService.resultEmotion(message, "✅", messageCode, args);
        return true;
    }

    protected boolean fail(MessageReceivedEvent message, String messageCode, Object... args) {
        commandsService.resultEmotion(message, "❌", messageCode, args);
        return false;
    }

    @Override
    public CatalystCommand getAnnotation() {
        if (annotation == null) {
            annotation = getClass().getDeclaredAnnotation(CatalystCommand.class);
        }
        return annotation;
    }

    protected Pair<String, Long> probeDuration(String input) {
        if (StringUtils.isBlank(input)) {
            return Pair.of(input, null);
        }
        String value = input.trim();
        String[] parts = value.split("\\s+");
        try {
            return Pair.of(value.substring(parts[0].length()).trim(), TimeSequenceParser.parseShort(parts[0]));
        } catch (Exception e) {
            return Pair.of(input, null);
        }
    }
}

File: Catalyst-Mono/Catalyst-Core/src/main/kotlin/net/kingchev/catalyst/ru/discord/command/model/AbstractCommand.kt
package net.kingchev.catalyst.ru.discord.command.model

import net.dv8tion.jda.api.events.interaction.command.SlashCommandInteractionEvent
import net.dv8tion.jda.api.events.message.MessageReceivedEvent
import net.dv8tion.jda.api.interactions.commands.build.SlashCommandData
import net.kingchev.catalyst.ru.core.config.CoreProperties
import net.kingchev.catalyst.ru.core.localezied.service.LocaleService
import net.kingchev.catalyst.ru.core.model.CommandStatus
import net.kingchev.catalyst.ru.core.service.GuildConfigService
import net.kingchev.catalyst.ru.core.service.UserConfigService
import net.kingchev.catalyst.ru.discord.command.service.InternalCommandService
import net.kingchev.catalyst.ru.discord.config.WorkerProperties
import net.kingchev.catalyst.ru.discord.context.model.MessageContext
import net.kingchev.catalyst.ru.discord.context.model.SlashContext
import net.kingchev.catalyst.ru.discord.message.service.MessageService
import org.springframework.beans.factory.annotation.Autowired

abstract class AbstractCommand : Command {
    @Autowired
    protected lateinit var internalCommandService: InternalCommandService

    @Autowired
    protected lateinit var localeService: LocaleService

    @Autowired
    protected lateinit var messageService: MessageService

    @Autowired
    protected lateinit var guildConfigService: GuildConfigService

    @Autowired
    protected lateinit var userConfigService: UserConfigService

    @Autowired
    protected lateinit var coreProperties: CoreProperties

    @Autowired
    protected lateinit var workerProperties: WorkerProperties

    open override fun isAvailable(event: MessageReceivedEvent, context: MessageContext): CommandStatus {
        val annotation = getAnnotation()
        val message = context.message
        val user = context.author
        val guild = context.guild
        val member = context.authorMember

        if (guild != null && annotation.dmOnly) {
            return internalCommandService.fail(context, CommandStatus.DM_ONLY_ERROR)
        }
        if (guild == null && annotation.guildOnly) {
            return internalCommandService.fail(context, CommandStatus.GUILD_ONLY_ERROR)
        }
        if (guild != null && !guild.selfMember.hasPermission(annotation.permissions.toList())) {
            return internalCommandService.fail(context, CommandStatus.BOT_PERMISSIONS_ERROR)
        }
        if (member != null && !context.authorMember.hasPermission(message.channel.asGuildMessageChannel(), annotation.userPermission.toList())) {
            return internalCommandService.fail(context, CommandStatus.USER_PERMISSIONS_ERROR)
        }
        return CommandStatus.SUCCESS
    }

    open override fun isAvailable(event: SlashCommandInteractionEvent, context: SlashContext): CommandStatus {
        val annotation = getAnnotation()
        val message = context.interaction
        val user = context.author
        val guild = context.guild
        val member = context.authorMember

        if (guild != null && annotation.dmOnly) {
            return internalCommandService.fail(context, CommandStatus.DM_ONLY_ERROR)
        }
        if (guild == null && annotation.guildOnly) {
            return internalCommandService.fail(context, CommandStatus.GUILD_ONLY_ERROR)
        }
        if (guild != null && !guild.selfMember.hasPermission(annotation.permissions.toList())) {
            return internalCommandService.fail(context, CommandStatus.BOT_PERMISSIONS_ERROR)
        }
        if (member != null && !context.authorMember.hasPermission(message.guildChannel, annotation.userPermission.toList())) {
            return internalCommandService.fail(context, CommandStatus.USER_PERMISSIONS_ERROR)
        }
        return CommandStatus.SUCCESS
    }
}
